Data/HashMap/Internal.hs

```diff
@@ -1771,37 +1756,156 @@ differenceWith f a b = foldlWithKey' go empty a
+    -- branch vs. branch
+    go s (BitmapIndexed b1 ary1) (BitmapIndexed b2 ary2) = intersectionArray s b1 b2 ary1 ary2
+    go s (BitmapIndexed b1 ary1) (Full ary2) = intersectionArray s b1 fullNodeMask ary1 ary2
+    go s (Full ary1) (BitmapIndexed b2 ary2) = intersectionArray s fullNodeMask b2 ary1 ary2
+    go s (Full ary1) (Full ary2) = intersectionArray s fullNodeMask fullNodeMask ary1 ary2
+    -- collision vs. branch
+    go s (BitmapIndexed b1 ary1) t2@(Collision h2 _ls2)
+      | b1 .&. m2 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        m2 = mask h2 s
+        i = sparseIndex b1 m2
+    go s t1@(Collision h1 _ls1) (BitmapIndexed b2 ary2)
+      | b2 .&. m1 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        m1 = mask h1 s
+        i = sparseIndex b2 m1
+    go s (Full ary1) t2@(Collision h2 _ls2) = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        i = index h2 s
+    go s t1@(Collision h1 _ls1) (Full ary2) = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        i = index h1 s
+
+    intersectionArray s b1 b2 ary1 ary2
+      -- don't create an array of size zero in intersectionArrayBy
+      | b1 .&. b2 == 0 = Empty
+      | otherwise = runST $ do
+        (b, len, ary) <- intersectionArrayBy (go (s + bitsPerSubkey)) b1 b2 ary1 ary2
+        case len of
+          0 -> pure Empty
+          1 -> A.read ary 0
+          _ -> bitmapIndexedOrFull b <$> (A.unsafeFreeze =<< A.shrink ary len)
+{-# INLINE intersectionWithKey# #-}
+
+intersectionArrayBy ::
+  ( HashMap k v1 ->
+    HashMap k v2 ->
+    HashMap k v3
+  ) ->
+  Bitmap ->
+  Bitmap ->
+  A.Array (HashMap k v1) ->
+  A.Array (HashMap k v2) ->
+  ST s (Bitmap, Int, A.MArray s (HashMap k v3))
+intersectionArrayBy f !b1 !b2 !ary1 !ary2 = do
+  mary <- A.new_ $ popCount bIntersect
+  -- iterate over nonzero bits of b1 .&. b2
+  let go !i !i1 !i2 !b !bFinal
+        | b == 0 = pure (i, bFinal)
+        | testBit $ b1 .&. b2 = do
+          x1 <- A.indexM ary1 i1
+          x2 <- A.indexM ary2 i2
+          case f x1 x2 of
+            Empty -> go i (i1 + 1) (i2 + 1) b' (bFinal .&. complement m)
+            _ -> do
+              A.write mary i $! f x1 x2
+              go (i + 1) (i1 + 1) (i2 + 1) b' bFinal
+        | testBit b1 = go i (i1 + 1) i2 b' bFinal
+        | otherwise = go i i1 (i2 + 1) b' bFinal
+        where
+          m = 1 `unsafeShiftL` countTrailingZeros b
+          testBit x = x .&. m /= 0
+          b' = b .&. complement m
+  (maryLen, bFinal) <- go 0 0 0 bCombined bIntersect
+  pure (bFinal, maryLen, mary)
+  where
+    bCombined = b1 .|. b2
+    bIntersect = b1 .&. b2
+{-# INLINE intersectionArrayBy #-}
+
+intersectionCollisions :: Eq k => (k -> v1 -> v2 -> (# v3 #)) -> A.Array (Leaf k v1) -> A.Array (Leaf k v2) -> ST s (Int, A.MArray s (Leaf k v3))
+intersectionCollisions f ary1 ary2 = do
+  mary2 <- A.thaw ary2 0 $ A.length ary2
```

Comment:
I wonder whether we actually need to allocate two arrays for this. The alternative would be to perform the search-and-swap operations on the output array itself.

It might be a bit tricky though – maybe leave it for a follow-up PR, so this one doesn't get too huge.

Reply:
I think the issue with this is that the type could change. For example if we have two arrays with the numbers as keys, and the arrays are both different types
`1 2 3 4`
`3 4 2 1`
Let's thaw the first array, and mutate it to
`(f 3 3) 2 1 4`
`f 3 3` could change the type to be something difference than the `2 1 4`.

Reply:
Ah, yes, good point. Unsafe coercions _might_ work for this, but I'd prefer not trying this in this PR.

Reply:
This is only an issue for `intersectionWithKey` and such; `intersection` itself has no type issue.

Reply:
There's another thing about `intersection` that's special: we can reuse the leaves.

Reply:
Yes, it would be better if `intersection` had custom code for handling collisions. Maybe this can be achieved by changing `intersectionWithKey#` to something similar to `filterMapAux`.

I'd slightly prefer if we'd leave this for a follow-up PR though.

Reply:
Here is the branch https://github.com/oberblastmeister/unordered-containers/tree/fast-intersection-sparseIndex

Reply:
I have recorded these ideas in https://github.com/haskell-unordered-containers/unordered-containers/issues/415.